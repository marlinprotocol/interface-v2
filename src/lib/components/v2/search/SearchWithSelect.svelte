<script lang="ts">
	import InputCard from '$lib/atoms/v2/cards/InputCard.svelte';
	import { inputClasses } from '$lib/atoms/v2/componentClasses';
	import Text from '$lib/atoms/texts/Text.svelte';
	import type { InputCardVariant } from '$lib/types/componentTypes';
	import Select from '$lib/components/v2/select/Select.svelte';
	import Button from '$lib/atoms/v2/buttons/Button.svelte';

	export let dataList: (string | number)[] = [];
	export let searchValue: string | number | undefined = '';
	export let setSearchValue: (value: string | number, exactMatch?: boolean) => any;
	export let onSearchClick: (() => void) | undefined = undefined;
	export let title: string;
	export let showTitle = true;
	export let onlyFilters = false;
	export let placeholder = 'Search';
	export let styleClass = '';
	export let cardVariant: InputCardVariant | undefined = 'primary';
	export let disabled = false;
	export let isTableFilter = false;
	export let textSuffix = '';
	export let handleClearFilters: () => void = () => {};
	export let clearButton = false;

	let suggestions: (string | number)[] = [];
	let showSuggestions = false;

	const handleSearch = async (event: Event) => {
		const input = event.target as HTMLInputElement;
		searchValue = input.value;
		// close suggestions if search is empty
		if (searchValue === '') {
			showSuggestions = false;
			suggestions = [];
			await setSearchValue(searchValue, false);
			return;
		}
		suggestions = dataList.filter((item) =>
			item
				.toString()
				.toLowerCase()
				.includes((searchValue ?? '').toString().toLowerCase())
		);
		showSuggestions = suggestions.length > 0;
		await setSearchValue(searchValue, false);
	};
</script>

<div class="flex flex-1 flex-col">
	<p class="mb-4 font-poppins text-base font-light text-[#030115]">{placeholder}</p>
	<InputCard {styleClass} variant={cardVariant}>
		<div class="flex flex-1">
			<div class="search-container flex flex-1">
				{#if showTitle}
					<div class="flex items-center justify-between">
						<div class="flex items-center gap-1">
							<Text variant="small" text={title} />
						</div>
					</div>
				{/if}
				<div class="flex flex-1 items-center">
					<input
						class="{isTableFilter
							? inputClasses.searchInputText
							: inputClasses.inputText} placeholder-[#A8A8A8]"
						{placeholder}
						value={searchValue ?? ''}
						on:input={handleSearch}
						disabled={onlyFilters || disabled}
						on:click={() => onSearchClick?.()}
					/>
					<Select
						{title}
						{dataList}
						bind:value={searchValue}
						setValue={(value) => setSearchValue(value, true)}
						bind:showSuggestions
						bind:suggestions
						{textSuffix}
					/>
				</div>
			</div>
			{#if clearButton && handleClearFilters}
				<Button
					variant="filled"
					size="medium"
					styleClass="w-[140px] ml-4"
					onclick={handleClearFilters}>CLEAR</Button
				>
			{/if}
		</div>
	</InputCard>
</div>
