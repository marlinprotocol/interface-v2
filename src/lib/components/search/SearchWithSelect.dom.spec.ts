import { describe, it, expect } from 'vitest';
import { render } from '@testing-library/svelte';
import SearchWithSelect from './SearchWithSelect.svelte';

describe('SearchWithSelect', () => {
	it("render's component properly", () => {
		const { container } = render(SearchWithSelect, {
			props: {
				placeholder: 'Search...',
				setSearchValue: vi.fn(),
				title: 'Test Title'
			}
		});

		expect(container).toMatchSnapshot();
	});

	it('renders correctly', () => {
		const { getByPlaceholderText } = render(SearchWithSelect, {
			props: {
				placeholder: 'Search...',
				setSearchValue: () => {},
				title: 'Test Title'
			}
		});

		const inputElement = getByPlaceholderText('Search...') as HTMLInputElement;
		expect(inputElement).toBeTruthy();
	});

	it('is disabled when disabled prop is true', () => {
		const { getByPlaceholderText } = render(SearchWithSelect, {
			props: {
				placeholder: 'Search...',
				disabled: true,
				setSearchValue: () => {},
				title: 'Test Title'
			}
		});

		const inputElement = getByPlaceholderText('Search...') as HTMLInputElement;
		expect(inputElement.disabled).toBe(true);
	});

	it('calls onSearchClick when search button is clicked', () => {
		const mockSetSearchValue = vi.fn();
		const { getByPlaceholderText } = render(SearchWithSelect, {
			props: {
				placeholder: 'Search...',
				setSearchValue: () => {},
				title: 'Test Title',
				onSearchClick: mockSetSearchValue
			}
		});

		const searchButton = getByPlaceholderText('Search...');
		searchButton.click();

		expect(mockSetSearchValue).toHaveBeenCalledOnce();
	});
});
