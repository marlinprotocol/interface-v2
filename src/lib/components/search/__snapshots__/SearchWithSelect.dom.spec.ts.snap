// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchWithSelect > render's component properly 1`] = `
<div>
  <div
    class="relative flex flex-1 flex-col"
  >
     
    <div
      class="w-full bg-base-100 px-6 py-[30px] rounded-3xl border border-grey-100"
      data-testid="input-card"
    >
      <div
        class="search-container"
      >
        <div
          class="flex items-center justify-between"
        >
          <div
            class="flex items-center gap-1"
          >
            <div
              class=" text-sm font-normal"
              data-testid="text"
            >
              Test Title
            </div>
          </div>
        </div>
         
        <div
          class="flex items-center"
        >
          <input
            class="input input-ghost h-[30px] w-full text-body p-0 font-normal text-grey-800 disabled:opacity-70 disabled:text-grey-800 disabled:placeholder:text-grey-800/[.2] focus-within:text-grey-800 placeholder:text-grey-800/[.2] focus:outline-none focus-within:border-b-2 focus:bg-transparent placeholder-[#A8A8A8]"
            placeholder="Search..."
          />
           
          <div
            class="search-container relative flex w-fit items-baseline svelte-1efctiy"
            data-testid="select"
            id="select-test title"
          >
            <div
              class="flex h-auto items-center svelte-1efctiy"
            >
               
              <button
                class="w-max"
                data-testid="collapse-button"
                disabled=""
                id=""
              >
                <img
                  alt="Collapse Icon"
                  class="icon-invert rotate-0 transform transition duration-200"
                  src="/images/chevronDownIcon.svg"
                />
              </button>
            </div>
             
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
